File: Data-Structures/graph.js
let q = new queue();
class graph{
    data;
    edges;
    constructor(data){
        this.data=data;
        this.edges=[];
    }
    addEdge(node){
        this.edges.push(node);
        node.edges.push(this);
    }
    print(){
        q.clear();
        //Using BFS;
        q.enqueue(this);
        let visited=[];
        while(!q.isEmpty()){
            let temp=q.dequeue();
            if(visited.includes(temp.data)){
                continue;
            }
            console.log(temp.data);
            visited.push(temp.data);
            for(let i=0;i<temp.edges.length;i++){
                q.enqueue(temp.edges[i]);
            }
        }
    }
    sum(){
        let sum=0;
        let visited=[];
        q.clear();
        q.enqueue(this);
        while(!q.isEmpty()){
            let temp=q.dequeue();
            if(visited.includes(temp.data)){
                continue;
            }
            sum+=temp.data;
            visited.push(temp.data);
            for(let i=0;i<temp.edges.length;i++){
                q.enqueue(temp.edges[i]);
            }
        }
        return sum;
    }
    max(){
        let max=-Infinity;
        let visited=[];
        q.clear();
        q.enqueue(this);
        while(!q.isEmpty()){
            let temp=q.dequeue();
            if(visited.includes(temp.data)){
                continue;
            }
            if(temp.data>max){
                max=temp.data;
            }
            visited.push(temp.data);
            for(let i=0;i<temp.edges.length;i++){
                q.enqueue(temp.edges[i]);
            }
        }
        return max;
    }
    min(){
        let min=Infinity;
        let visited=[];
        q.clear();
        q.enqueue(this);
        while(!q.isEmpty()){
            let temp=q.dequeue();
            if(visited.includes(temp.data)){
                continue;
            }
            if(temp.data<min){
                min=temp.data;
            }
            visited.push(temp.data);
            for(let i=0;i<temp.edges.length;i++){
                q.enqueue(temp.edges[i]);
            }
        }
        return min;
    }
    count(){
        let count=0;
        let visited=[];
        q.clear();
        q.enqueue(this);
        while(!q.isEmpty()){
            let temp=q.dequeue();
            if(visited.includes(temp.data)){
                continue;
            }
            count++;
            visited.push(temp.data);
            for(let i=0;i<temp.edges.length;i++){
                q.enqueue(temp.edges[i]);
            }
        }
        return count;
    }
    average(){
        return this.sum()/this.count();
    }
    sort(){
        let visited=[];
        q.clear();
        q.enqueue(this);
        let arr=[];
        while(!q.isEmpty()){
            let temp=q.dequeue();
            if(visited.includes(temp.data)){
                continue;
            }
            arr.push(temp.data);
            visited.push(temp.data);
            for(let i=0;i<temp.edges.length;i++){
                q.enqueue(temp.edges[i]);
            }
        }
        arr.sort();
        return arr;
    }
}
module.exports=graph;

File: Data-Structures/graph.cpp
#include <iostream> 
#include <list>

using namespace std;
class Graph{
    int v;
    list<int> *adj;
    public:
      Graph(int v);
      void addEDGE(int a,int b);
       void BFS(int s); 
};
Graph::Graph(int v){
    this->v=v;
    adj=new list<int>[v];
};
void Graph::addEDGE(int a,int b){
   adj[a].push_back(b);
};
void Graph::BFS(int s){
     bool *visited = new bool[v];
     for(int i=0;i<v;i++){
         visited[i]=false;
     }
     list<int> queue;
     visited[s]=true;
     queue.push_back(s);
     list<int>::iterator i;
      while(!queue.empty()){
          s = queue.front();
        cout << s << " ";
        queue.pop_front();
        for (i = adj[s].begin(); i != adj[s].end(); ++i)
        {
            if (!visited[*i])
            {
                visited[*i] = true;
                queue.push_back(*i);
            }
        }
}
};
int main()
{
    Graph g(4);
    g.addEDGE(0, 1);
    g.addEDGE(0, 2);
    g.addEDGE(1, 2);
    g.addEDGE(2, 0);
    g.addEDGE(2, 3);
    g.addEDGE(3, 3);
    g.BFS(2);
 
    return 0;
}
